File: U1/TI1.4 RESTFULL_DOTNET_SINBDD_GR01/01 SERVIDOR/ConUni_Restfull_Dotnet_GR01/ConUni_Restfull_Dotnet_GR01/ec/edu/monster/ws/MasaController.cs
using Microsoft.AspNetCore.Http;
using Microsoft.AspNetCore.Mvc;
using ConUni_Restfull_Dotnet_GR01.ec.edu.monster.services;
using ConUni_Restfull_Dotnet_GR01.ec.edu.monster.models;
using ConUni_Restfull_Dotnet_GR01.ec.edu.monster.constants;
using ConUni_Restfull_Dotnet_GR01.ec.edu.monster.validators;

namespace ConUni_Restfull_Dotnet_GR01.ec.edu.monster.ws
{
    /// <summary>
    /// Controlador REST para conversiones de masa
    /// Soporta conversiones entre: Kilogramo, Quintal, Libra
    /// </summary>
    [Route("api/[controller]")]
    [ApiController]
    [Produces("application/json")]
    public class MasaController : ControllerBase
    {
        private readonly MasaService _masaService;
        
        // Unidades válidas para masa
  private readonly HashSet<string> _unidadesValidas = new HashSet<string>(StringComparer.OrdinalIgnoreCase)
      {
          "kilogramo", "quintal", "libra"
      };

        public MasaController(MasaService masaService)
        {
      _masaService = masaService;
     }

        /// <summary>
        /// Convierte valores entre diferentes unidades de masa
        /// </summary>
        /// <param name="request">Solicitud de conversión con valor, unidad origen y unidad destino</param>
        /// <returns>Resultado de la conversión con valores exactos y redondeados</returns>
/// <response code="200">Conversión exitosa o error de validación</response>
      /// <remarks>
        /// Unidades soportadas: Kilogramo, Quintal, Libra (case-insensitive)
        ///
        /// Ejemplo de solicitud:
        ///
     ///     POST /api/Masa/convertir
      ///     {
        ///        "valor": "100",
  ///        "unidadOrigen": "Kilogramo",
 ///         "unidadDestino": "Libra"
   ///     }
        ///
        /// </remarks>
 [HttpPost("convertir")]
        [ProducesResponseType(typeof(ConversionResultModel), StatusCodes.Status200OK)]
        [ProducesResponseType(typeof(ConversionResultModel), StatusCodes.Status400BadRequest)]
     public ActionResult<ConversionResultModel> Convertir([FromBody] ConversionRequest request)
{
            // Normalizar las unidades a minúsculas para comparación
       var origen = request.UnidadOrigen.Trim().ToLower();
  var destino = request.UnidadDestino.Trim().ToLower();

    // Primero validar que ambas unidades sean soportadas
            if (!_unidadesValidas.Contains(origen) || !_unidadesValidas.Contains(destino))
            {
        return Ok(ConversionResultModel.Fallo(
         new ConversionErrorModel(
     ErrorConstants.ERROR_CONVERSION_MASA,
    $"Conversión de {request.UnidadOrigen} a {request.UnidadDestino} no está soportada",
ErrorConstants.TIPO_CONVERSION,
TryParseDouble(request.Valor),
             request.UnidadOrigen,
    $"Las unidades soportadas son: {MasaConstants.KILOGRAMO}, {MasaConstants.QUINTAL}, {MasaConstants.LIBRA}"
            )
                ));
       }

        // Determinar qué conversión realizar basándose en origen y destino
     var resultado = (origen, destino) switch
    {
                ("kilogramo", "quintal") => _masaService.ConvertirKilogramoAQuintal(request.Valor),
        ("quintal", "kilogramo") => _masaService.ConvertirQuintalAKilogramo(request.Valor),
  ("kilogramo", "libra") => _masaService.ConvertirKilogramoALibra(request.Valor),
      ("libra", "kilogramo") => _masaService.ConvertirLibraAKilogramo(request.Valor),
             ("quintal", "libra") => _masaService.ConvertirQuintalALibra(request.Valor),
      ("libra", "quintal") => _masaService.ConvertirLibraAQuintal(request.Valor),

  // Conversión de una unidad a sí misma (solo llega aquí si ambas unidades son válidas)
       _ when origen == destino => HandleSameUnitConversion(request.Valor, request.UnidadOrigen, request.UnidadDestino),

 // Esta línea nunca debería ejecutarse debido a la validación anterior
      _ => ConversionResultModel.Fallo(
        new ConversionErrorModel(
       ErrorConstants.ERROR_CONVERSION_MASA,
                $"Conversión de {request.UnidadOrigen} a {request.UnidadDestino} no está soportada",
  ErrorConstants.TIPO_CONVERSION,
      TryParseDouble(request.Valor),
    request.UnidadOrigen,
           $"Las unidades soportadas son: {MasaConstants.KILOGRAMO}, {MasaConstants.QUINTAL}, {MasaConstants.LIBRA}"
          )
            )
   };

    return Ok(resultado);
     }

        /// <summary>
        /// Maneja conversiones donde origen y destino son la misma unidad
        /// </summary>
private ConversionResultModel HandleSameUnitConversion(string valorString, string unidadOrigen, string unidadDestino)
        {
    // Primero validar que el valor sea válido
            var error = BaseValidator.ValidarStringPositivo(valorString, unidadOrigen, out double valor);
        if (error != null)
   return ConversionResultModel.Fallo(error);

         return ConversionResultModel.Exito(
         new UnidadConversionModel(
         valor,
       valor,
 CapitalizarUnidad(unidadOrigen),
         CapitalizarUnidad(unidadDestino),
    "Masa",
       1.0
    )
    );
     }

     /// <summary>
     /// Intenta convertir string a double para casos de error
   /// </summary>
private double? TryParseDouble(string value)
        {
            if (double.TryParse(value, out double result))
              return result;
            return null;
        }

 /// <summary>
     /// Capitaliza el nombre de la unidad (Primera letra en mayúscula)
    /// </summary>
  private string CapitalizarUnidad(string unidad)
        {
       if (string.IsNullOrWhiteSpace(unidad))
     return string.Empty;

       unidad = unidad.Trim().ToLower();
 return char.ToUpper(unidad[0]) + unidad.Substring(1);
        }
    }
}

File: U1/TI1.3 RESTFULL_JAVA_SINBDD_GR01/01 SERVIDOR/ConUni_Restfull_Java_GR01/src/java/ec/edu/monster/ws/MasaController.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/WebServices/GenericResource.java to edit this template
 */
package ec.edu.monster.ws;

import ec.edu.monster.constants.ConversionConstants.Masa;
import ec.edu.monster.models.ConversionRequest;
import ec.edu.monster.models.ConversionResultModel;
import ec.edu.monster.services.MasaService;
import jakarta.inject.Inject;
import jakarta.ws.rs.Consumes;
import jakarta.ws.rs.POST;
import jakarta.ws.rs.Path;
import jakarta.ws.rs.Produces;
import jakarta.ws.rs.core.MediaType;
import jakarta.ws.rs.core.Response;

/**
 * REST Web Service para conversiones de masa
 * Endpoints disponibles (todos POST con JSON):
 * - /api/masa - Conversión general (basada en unidadOrigen y unidadDestino)
 *
 * Unidades soportadas: Kilogramo, Quintal, Libra
 *
 * Ejemplo de request:
 * {
 *   "valor": "100",
 *   "unidadOrigen": "Kilogramo",
 *   "unidadDestino": "Libra"
 * }
 *
 * @author jeffe
 */
@Path("masa")
public class MasaController {

    @Inject
    private MasaService masaService;

    /**
     * Endpoint POST para convertir entre unidades de masa
     * @param request ConversionRequest con valor, unidadOrigen y unidadDestino
     * @return Response con ConversionResultModel en JSON
     */
    @POST
    @Consumes(MediaType.APPLICATION_JSON)
    @Produces(MediaType.APPLICATION_JSON)
    public Response convertir(ConversionRequest request) {
        ConversionResultModel resultado;

        String origen = request.getUnidadOrigen();
        String destino = request.getUnidadDestino();
        String valor = request.getValor();

        // Determinar qué conversión realizar basándose en origen y destino
        if (Masa.KILOGRAMO.equalsIgnoreCase(origen) && Masa.QUINTAL.equalsIgnoreCase(destino)) {
            resultado = masaService.convertirKilogramoAQuintal(valor);
        } else if (Masa.QUINTAL.equalsIgnoreCase(origen) && Masa.KILOGRAMO.equalsIgnoreCase(destino)) {
            resultado = masaService.convertirQuintalAKilogramo(valor);
        } else if (Masa.KILOGRAMO.equalsIgnoreCase(origen) && Masa.LIBRA.equalsIgnoreCase(destino)) {
            resultado = masaService.convertirKilogramoALibra(valor);
        } else if (Masa.LIBRA.equalsIgnoreCase(origen) && Masa.KILOGRAMO.equalsIgnoreCase(destino)) {
            resultado = masaService.convertirLibraAKilogramo(valor);
        } else if (Masa.QUINTAL.equalsIgnoreCase(origen) && Masa.LIBRA.equalsIgnoreCase(destino)) {
            resultado = masaService.convertirQuintalALibra(valor);
        } else if (Masa.LIBRA.equalsIgnoreCase(origen) && Masa.QUINTAL.equalsIgnoreCase(destino)) {
            resultado = masaService.convertirLibraAQuintal(valor);
        } else {
            // Conversión no soportada
            return Response
                .status(Response.Status.BAD_REQUEST)
                .entity("{\"error\": \"Conversión no soportada. Unidades válidas: Kilogramo, Quintal, Libra\"}")
                .build();
        }

        // Retornar respuesta según el resultado
        if (resultado.isExitoso()) {
            return Response.ok(resultado).build();
        } else {
            return Response
                .status(Response.Status.BAD_REQUEST)
                .entity(resultado)
                .build();
        }
    }
}
